# Exploratory Data Analysis
## K-Pop Database (1992-2020)
Public dataset obtained from [here](https://www.kaggle.com/kimjihoo/kpopdb).

A high level exploration into a K-Pop Database between 1992 and 2020. I will explore the idols, groups, and music videos release between these years to determine what insights can be obtained. 

## Imports

### Libraries

In [1]:
import pandas as pd
import plotly.express as px

### Data

Our file imports are in the following structure:

ea_kpop_database_files/ <br>
├── kpop_idols.csv <br>
├── kpop_idols_boy_groups.csv <br>
├── kpop_idols_girl_groups.csv <br>
└── kpop_music_videos.csv <br>

Import csv files to dataframes and make copies to protect original imports.

In [2]:
df_idols_import = pd.read_csv("eda_kpop_database_files/kpop_idols.csv")
df_boy_groups_import = pd.read_csv("eda_kpop_database_files/kpop_idols_boy_groups.csv")
df_girl_groups_import = pd.read_csv("eda_kpop_database_files/kpop_idols_girl_groups.csv")
df_music_videos_import = pd.read_csv("eda_kpop_database_files/kpop_music_videos.csv")

df_idols = df_idols_import.copy()
df_boy_groups = df_boy_groups_import.copy()
df_girl_groups = df_girl_groups_import.copy()
df_music_videos = df_music_videos_import.copy()

Preview our main dataframe.

In [3]:
df_idols

Stage Name      Full Name Korean Name  ... Birthplace Other Group Gender
0           A.M  Seong Hyunwoo         성현우  ...        NaN         NaN      M
1           Ace  Jang Wooyoung         장우영  ...        NaN         NaN      M
2          Aeji      Kwon Aeji         권애지  ...      Daegu         NaN      F
3          AhIn       Lee Ahin         이아인  ...      Wonju         NaN      F
4          Ahra        Go Ahra         고아라  ...      Yeosu         NaN      F
...         ...            ...         ...  ...        ...         ...    ...
1305        Ziu    Park Heejun         박희준  ...        NaN         NaN      M
1306         ZN      Bae Jinye         배진예  ...    Bucheon       UNI.T      F
1307        Zoa     Cho Hyewon         조혜원  ...        NaN         NaN      F
1308       Zuho       Bae Juho         백주호  ...        NaN         NaN      M
1309       Zuny      Kim Joomi         김주미  ...    Gwangju         NaN      F

[1310 rows x 10 columns]

# Exploratory Data Analysis
A high level version.


## Data Wrangling

The main purpose here is to see what type of data we are working with. 

Is any of the data dirty, or require modifications? (i.e. datetime)

A look into the datatypes for each column in our dataframes. 

Both group dataframes contain the same structure and types.

In [4]:
df_idols.dtypes

Stage Name       object
Full Name        object
Korean Name      object
K. Stage Name    object
Date of Birth    object
Group            object
Country          object
Birthplace       object
Other Group      object
Gender           object
dtype: object

In [5]:
df_boy_groups.dtypes

Name            object
Short           object
Korean Name     object
Debut           object
Company         object
Members          int64
Orig. Memb.      int64
Fanclub Name    object
Active          object
dtype: object

In [6]:
df_music_videos.dtypes

Date           object
Artist         object
Song Name      object
Korean Name    object
Director       object
Video          object
Type           object
Release        object
dtype: object

Some initial thoughts:

* The date columns are in Strings and will need to be coverted to a datetime format
* Numeric columns are already set properly
* The Idol and Group dataframes can be merged
* Column headers should be renamed entirely for consistency


Merging both girl and boy group dataframes, whilst resetting the index.

In [7]:
frames_to_concat = [df_boy_groups, df_girl_groups]
df_groups = pd.concat(frames_to_concat, ignore_index=True, sort=False)

In [8]:
df_groups

Name         Short Korean Name  ... Orig. Memb. Fanclub Name  Active
0            100%           NaN        백퍼센트  ...           7   Perfection     Yes
1             14U           NaN         원포유  ...          14          NaN     Yes
2           1the9           NaN        원더나인  ...           9          NaN     Yes
3             24K           NaN        투포케이  ...           6          24U     Yes
4             2AM           NaN        투에이엠  ...           4         I Am      No
..            ...           ...         ...  ...         ...          ...     ...
294      We Girls           NaN         위걸스  ...           8          NaN     Yes
295     Weki Meki          WeMe        위키미키  ...           8          NaN     Yes
296          WJSN  Cosmic Girls        우주소녀  ...          12        Ujung     Yes
297  Wonder Girls            WG        원더걸스  ...           5    Wonderful      No
298    Yellow Bee           NaN         옐로비  ...           5          NaN     Yes

[299 rows x 9 columns]

Beginning the column header rename process. Using a dictionary will give full control here.

In [9]:
idols_columnHeader_rename_dict = {
  'Stage Name': 'idol_stage_name',
  'Full Name': 'idol_full_name',
  'Korean Name': 'idol_korean_name',
  'K. Stage Name': 'idol_korean_stage_name',
  'Date of Birth': 'idol_date_of_birth',
  'Group': 'idol_group_name',
  'Country': 'idol_country',
  'Birthplace': 'idol_birthplace',
  'Other Group': 'idol_other_group_name',
  'Gender': 'idol_gender'
  }

print("\nBefore rename")
display(df_idols.dtypes)

# call rename() method
df_idols.rename(columns=idols_columnHeader_rename_dict,
          inplace=True)
 
print("\nAfter rename")
display(df_idols.dtypes)


Before rename


Stage Name       object
Full Name        object
Korean Name      object
K. Stage Name    object
Date of Birth    object
Group            object
Country          object
Birthplace       object
Other Group      object
Gender           object
dtype: object


After rename


idol_stage_name           object
idol_full_name            object
idol_korean_name          object
idol_korean_stage_name    object
idol_date_of_birth        object
idol_group_name           object
idol_country              object
idol_birthplace           object
idol_other_group_name     object
idol_gender               object
dtype: object

Repeat rename process for remaining dataframes.

In [10]:
# rename the group dataframes.
groups_columnHeader_rename_dict = {
  'Name': 'group_name',
  'Short': 'group_name_short',
  'Korean Name': 'group_korean_name',
  'Debut': 'group_debut',
  'Company': 'group_company',
  'Members': 'group_members',
  'Orig. Memb.': 'group_original_members',
  'Fanclub Name': 'group_fanclub_name',
  'Active': 'group_active'
  }

# rename the music videos dataframe.
music_videos_columnHeader_rename_dict = {
  'Date': 'mv_date',
  'Artist': 'mv_artist',
  'Song Name': 'mv_song_name',
  'Korean Name': 'mv_korean_name',
  'Director': 'mv_director',
  'Video': 'mv_video_url',
  'Type': 'mv_type',
  'Release': 'mv_release'
  }
 
# call rename() method
df_groups.rename(columns=groups_columnHeader_rename_dict,
          inplace=True)

df_music_videos.rename(columns=music_videos_columnHeader_rename_dict,
          inplace=True)

# quick sanity check
df_groups.dtypes

group_name                object
group_name_short          object
group_korean_name         object
group_debut               object
group_company             object
group_members              int64
group_original_members     int64
group_fanclub_name        object
group_active              object
dtype: object

In [11]:
df_music_videos.dtypes

mv_date           object
mv_artist         object
mv_song_name      object
mv_korean_name    object
mv_director       object
mv_video_url      object
mv_type           object
mv_release        object
dtype: object

Merging the Idol and Groups dataframe with a Left Outer Join. I want to preserve all data in the Idol (left) dataframe. 

In [12]:
df = pd.merge(df_idols, df_groups, left_on="idol_group_name", right_on="group_name", how='left')
df

idol_stage_name idol_full_name  ... group_fanclub_name group_active
0                A.M  Seong Hyunwoo  ...                NaN          Yes
1                Ace  Jang Wooyoung  ...                NaN          Yes
2               Aeji      Kwon Aeji  ...                NaN          NaN
3               AhIn       Lee Ahin  ...                NaN          NaN
4               Ahra        Go Ahra  ...                NaN          Yes
...              ...            ...  ...                ...          ...
1305             Ziu    Park Heejun  ...                NaN          Yes
1306              ZN      Bae Jinye  ...                NaN          NaN
1307             Zoa     Cho Hyewon  ...                NaN          NaN
1308            Zuho       Bae Juho  ...            Fantasy          Yes
1309            Zuny      Kim Joomi  ...             Lavely       Hiatus

[1310 rows x 19 columns]

Converting date columns from Strings to valid Datetime objects.

In [13]:
df[["idol_date_of_birth", "group_debut"]] = df[["idol_date_of_birth", "group_debut"]].apply(pd.to_datetime)
df.dtypes

idol_stage_name                   object
idol_full_name                    object
idol_korean_name                  object
idol_korean_stage_name            object
idol_date_of_birth        datetime64[ns]
idol_group_name                   object
idol_country                      object
idol_birthplace                   object
idol_other_group_name             object
idol_gender                       object
group_name                        object
group_name_short                  object
group_korean_name                 object
group_debut               datetime64[ns]
group_company                     object
group_members                    float64
group_original_members           float64
group_fanclub_name                object
group_active                      object
dtype: object

Quick preview to confirm update:

In [14]:
df

idol_stage_name idol_full_name  ... group_fanclub_name group_active
0                A.M  Seong Hyunwoo  ...                NaN          Yes
1                Ace  Jang Wooyoung  ...                NaN          Yes
2               Aeji      Kwon Aeji  ...                NaN          NaN
3               AhIn       Lee Ahin  ...                NaN          NaN
4               Ahra        Go Ahra  ...                NaN          Yes
...              ...            ...  ...                ...          ...
1305             Ziu    Park Heejun  ...                NaN          Yes
1306              ZN      Bae Jinye  ...                NaN          NaN
1307             Zoa     Cho Hyewon  ...                NaN          NaN
1308            Zuho       Bae Juho  ...            Fantasy          Yes
1309            Zuny      Kim Joomi  ...             Lavely       Hiatus

[1310 rows x 19 columns]

To aid vizualizations later, adding assisting numeric columns which will be the current year - the date year.

In [15]:
now = pd.to_datetime("now")
now

Timestamp('2022-02-06 20:58:59.268448')

The month is also added into the calcuations for more accuracy. 

In [16]:
df["idol_date_of_birth_age"] = (now.year - df["idol_date_of_birth"].dt.year) - ((now.month - df["idol_date_of_birth"].dt.month) < 0)
df["group_debut_age"] = (now.year - df["group_debut"].dt.year) - ((now.month - df["group_debut"].dt.month) < 0)
df[["idol_date_of_birth", "idol_date_of_birth_age", "group_debut", "group_debut_age"]]

idol_date_of_birth  idol_date_of_birth_age group_debut  group_debut_age
0            1996-12-31                      25  2019-07-09              2.0
1            1992-08-28                      29  2015-10-31              6.0
2            1999-10-25                      22         NaT              NaN
3            1999-09-27                      22         NaT              NaN
4            2001-02-21                      21  2017-07-05              4.0
...                 ...                     ...         ...              ...
1305         1997-06-16                      24  2015-10-31              6.0
1306         1994-06-09                      27         NaT              NaN
1307         2005-05-31                      16         NaT              NaN
1308         1996-07-04                      25  2016-10-05              5.0
1309         1994-12-08                      27  2013-03-07              8.0

[1310 rows x 4 columns]

Confirm the updated data types.

In [17]:
df.dtypes

idol_stage_name                   object
idol_full_name                    object
idol_korean_name                  object
idol_korean_stage_name            object
idol_date_of_birth        datetime64[ns]
idol_group_name                   object
idol_country                      object
idol_birthplace                   object
idol_other_group_name             object
idol_gender                       object
group_name                        object
group_name_short                  object
group_korean_name                 object
group_debut               datetime64[ns]
group_company                     object
group_members                    float64
group_original_members           float64
group_fanclub_name                object
group_active                      object
idol_date_of_birth_age             int64
group_debut_age                  float64
dtype: object

## Exploration

Using Plotly Express, the idea here is to create data visualizations and gather insights.

Pandas functions can also be used to gather fast information that would other be overkil for a data viz.

I use the describe() function to gather the counts and mean values of our numeric columns.

In [18]:
df.describe()

group_members  ...  group_debut_age
count     846.000000  ...       846.000000
mean        6.839243  ...         6.601655
std         2.966507  ...         3.564579
min         2.000000  ...         1.000000
25%         5.000000  ...         4.000000
50%         6.000000  ...         6.000000
75%         8.000000  ...         9.000000
max        18.000000  ...        16.000000

[8 rows x 4 columns]

**Insight:**
* 846 idols out of 1309 total idols belong to a K-Pop group
* The mean number of members per group is 6.8
* On average, groups lose members over time with a mean score of 7.1
* The mean age of an idol is 25.6 years in February 2022. The oldest being 44 and youngest being 16.
* Most groups debuted 6.6 years ago

In [19]:
df["idol_gender"].value_counts()

M    676
F    634
Name: idol_gender, dtype: int64

**Insight:** There are currently more male idols that females at 51.6%.


In [20]:
df["idol_group_name"].isna().sum()

91

**Insight:** There are only 91 solo idols which represents 6.94% of the overall total.

Breakdown of groups by Companies that they work under.

In [21]:
df_company_counts = df.groupby(['group_company', 'idol_gender']).size().reset_index(name='counts')

fig = px.bar(df_company_counts, x='group_company', y='counts', color="idol_gender").update_xaxes(categoryorder="total descending")
fig

**Insight:**
* SM is the largest Company followed closely by JYP
* The larger companies appear to hire more male groups
* Some companies appear to only hire groups of a certain gender

Checking one of the male-only companies to validate the data.

In [22]:
df_company_counts[df_company_counts['group_company'].str.contains("Big Hit")]

group_company idol_gender  counts
9       Big Hit           M      12

In [23]:
df_groups[df_groups["group_name"] == "GLAM"]

group_name group_name_short  ... group_fanclub_name group_active
219       GLAM              NaN  ...                NaN           No

[1 rows x 9 columns]

In [24]:
df[df["idol_group_name"] == "GLAM"]

Empty DataFrame
Columns: [idol_stage_name, idol_full_name, idol_korean_name, idol_korean_stage_name, idol_date_of_birth, idol_group_name, idol_country, idol_birthplace, idol_other_group_name, idol_gender, group_name, group_name_short, group_korean_name, group_debut, group_company, group_members, group_original_members, group_fanclub_name, group_active, idol_date_of_birth_age, group_debut_age]
Index: []

**Problem:** Whilst a group called "GLAM" goes exist under the company "Big Hit", there were no matching idols when both dataframes were merged using the group name. The dataset is incomplete.


Checking the most common idol name:

In [25]:
df["idol_full_name"].value_counts().nlargest(10)

Lee Minhyuk      3
Lee Seunghyun    3
Lee Seoyoung     3
Kim Sohee        3
Kim Minseok      3
Kim Chaewon      3
Kim Jiwon        3
Kim Donghyun     3
Park Sooyoung    3
Kim Jihun        2
Name: idol_full_name, dtype: int64

A histogram to compare idol age and gender:

In [26]:
fig = px.histogram(df, x="idol_date_of_birth_age", color="idol_gender")
fig


**Insight:** Both male and female idols trend the same, with similar counts, when comparing ages.

Comparing group debut age, and their current activity status:

In [27]:
df_group_debut_age = df[df["group_debut_age"].notna()]
fig = px.histogram(df_group_debut_age, x="group_debut_age", color="group_active")
fig


**Insight:**
* As expected, groups with more recent debuts are active and skew as age increases
* There are still a fair number of active groups that debuted over 9 years ago

Using the dataprep library to convert full country names to ISO country codes (i.e. South Korea to KOR).

Credit: [Stack Overflow](https://stackoverflow.com/a/66309343)

In [28]:
!pip install dataprep
from dataprep.clean import clean_country

Adding a new column to our main dataframe with the ISO country code.

The ISO country code will be in Alpha-3 format which is used by the Plotly Express maps. 

In [29]:
df = clean_country(df, 'idol_country', output_format='alpha-3')
df

Country Cleaning Report:
	1296 values cleaned (98.93%)
Result contains 1310 (100.0%) values in the correct format and 0 null values (0.0%)


idol_stage_name idol_full_name  ... group_debut_age idol_country_clean
0                A.M  Seong Hyunwoo  ...             2.0                KOR
1                Ace  Jang Wooyoung  ...             6.0                KOR
2               Aeji      Kwon Aeji  ...             NaN                KOR
3               AhIn       Lee Ahin  ...             NaN                KOR
4               Ahra        Go Ahra  ...             4.0                KOR
...              ...            ...  ...             ...                ...
1305             Ziu    Park Heejun  ...             6.0                KOR
1306              ZN      Bae Jinye  ...             NaN                KOR
1307             Zoa     Cho Hyewon  ...             NaN                KOR
1308            Zuho       Bae Juho  ...             5.0                KOR
1309            Zuny      Kim Joomi  ...             8.0                KOR

[1310 rows x 22 columns]

Checking which country that idols belong to:

In [30]:
df_country_counts = df.groupby(['idol_country', 'idol_country_clean', 'idol_gender']).size().reset_index(name='counts')
df_country_counts

idol_country idol_country_clean idol_gender  counts
0     Australia                AUS           M       1
1        Canada                CAN           F       1
2        Canada                CAN           M       4
3         China                CHN           F      15
4         China                CHN           M      24
5     Hong Kong                HKG           F       2
6     Hong Kong                HKG           M       2
7     Indonesia                IDN           F       1
8     Indonesia                IDN           M       1
9         Japan                JPN           F      18
10        Japan                JPN           M       9
11     Malaysia                MYS           M       1
12  Philippines                PHL           F       1
13  South Korea                KOR           F     583
14  South Korea                KOR           M     621
15       Taiwan                TWN           F       5
16       Taiwan                TWN           M       1
17     Thailand                THA           F       3
18     Thailand                THA           M       3
19          USA                USA           F       5
20          USA                USA           M       9

**Insight:** We expected most idols to originate from South Korea

Removing South Korean idols from the dataframe to get a better visualization of the other countries:

In [31]:
df_country_counts_exludeSK = df_country_counts[df_country_counts["idol_country"] != "South Korea"]
fig = px.bar(df_country_counts_exludeSK, x='idol_country', y='counts', color="idol_gender").update_xaxes(categoryorder="total descending")
fig

**Insight:** 
* After South Korea, China and Japan produce more idols
* China, USA, and Canada are skewed towards male idols
* Japan and Taiwan are skewed towards female idols

Checking the idol countries again, over time by the year that they debuted: 

In [32]:
df_groups_country_counts = df[df["group_debut_age"].notna() & (df["idol_country"] != "South Korea")]
df_groups_country_counts["group_debut"] = df_groups_country_counts["group_debut"].dt.year
df_groups_country_counts = df_groups_country_counts.groupby(['group_debut', 'idol_country_clean', 'idol_country']).size().reset_index(name='counts')

df_groups_country_counts

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



group_debut idol_country_clean idol_country  counts
0          2008                THA     Thailand       1
1          2008                USA          USA       2
2          2009                CHN        China       1
3          2009                USA          USA       1
4          2010                CHN        China       2
5          2011                USA          USA       1
6          2012                CHN        China       4
7          2012                JPN        Japan       1
8          2014                CHN        China       3
9          2014                HKG    Hong Kong       1
10         2014                THA     Thailand       1
11         2014                USA          USA       2
12         2015                CHN        China       3
13         2015                HKG    Hong Kong       1
14         2015                THA     Thailand       1
15         2015                USA          USA       2
16         2016                CAN       Canada       1
17         2016                CHN        China       6
18         2016                HKG    Hong Kong       1
19         2016                JPN        Japan       2
20         2016                THA     Thailand       2
21         2016                USA          USA       1
22         2017                CAN       Canada       1
23         2017                CHN        China       2
24         2017                IDN    Indonesia       1
25         2017                JPN        Japan       3
26         2017                MYS     Malaysia       1
27         2017                TWN       Taiwan       1
28         2017                USA          USA       2
29         2018                AUS    Australia       1
30         2018                CHN        China       1
31         2018                JPN        Japan       7
32         2018                THA     Thailand       1
33         2018                TWN       Taiwan       2
34         2019                CHN        China       1
35         2019                JPN        Japan       2
36         2019                USA          USA       1
37         2020                USA          USA       1

Plotting this data to a timeseries map:

In [33]:
fig = px.scatter_geo(df_groups_country_counts, locations="idol_country_clean",
                     hover_name="idol_country", size="counts", size_max=60,
                    animation_frame="group_debut", 
                     projection="natural earth")

fig

**Insight:** 
* In 2016 we begin to see a higher influx of international idols
* In 2018 there is a surge of idols from Japan

Checking which cities most South Korean idols are from:

In [34]:
# SK by Birth Place
df_SKBirthplace_counts = df[df["idol_birthplace"].notna() & (df["idol_country"] == "South Korea")]
df_SKBirthplace_counts = df_SKBirthplace_counts.groupby(['idol_birthplace', 'idol_gender']).size().reset_index(name='counts')

fig = px.bar(df_SKBirthplace_counts, x='idol_birthplace', y='counts', color="idol_gender").update_xaxes(categoryorder="total descending")
fig

**Insight:** Most South Korean idols are from the capital, Seoul.

Checking which month of the year that most groups debuted in:

In [35]:
df_groups_debut_count = df[df["group_debut_age"].notna()]
df_groups_debut_count["group_debut"] = df_groups_debut_count["group_debut"].dt.month
df_groups_debut_count = df_groups_debut_count.groupby(['group_debut', 'idol_gender']).size().reset_index(name='counts')

import calendar
df_groups_debut_count['group_debut_month'] = df_groups_debut_count['group_debut'].apply(lambda x: calendar.month_abbr[x])
df_groups_debut_count

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



group_debut idol_gender  counts group_debut_month
0             1           F      43               Jan
1             1           M      32               Jan
2             2           F      25               Feb
3             2           M      15               Feb
4             3           F      42               Mar
5             3           M      30               Mar
6             4           F      20               Apr
7             4           M      81               Apr
8             5           F      32               May
9             5           M      63               May
10            6           F      28               Jun
11            6           M      23               Jun
12            7           F      50               Jul
13            7           M      37               Jul
14            8           F      60               Aug
15            8           M      42               Aug
16            9           F      15               Sep
17            9           M      42               Sep
18           10           F      30               Oct
19           10           M      72               Oct
20           11           F      26               Nov
21           11           M      16               Nov
22           12           F      10               Dec
23           12           M      12               Dec

In [36]:
fig = px.bar(df_groups_debut_count, x='group_debut_month', y='counts', color="idol_gender").update_xaxes(categoryorder='array', categoryarray= ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'])
fig

**Insight:** 
* Most groups debut in the April, May, August, and October months
* February and December are unpopular months for group debuts

Idols typically have a comeback twice per year. This can occur as a Summer Song during April and May, with a later comeback during October for award shows [[source]](https://www.reddit.com/r/kpophelp/comments/m3py4g/does_someone_know_when_are_there_usually_more/).

Checking how many members leave a group over time:

In [37]:
df_groups_members = df[df["idol_group_name"].notna() & (df["group_members"].notna()) & (df["group_original_members"].notna())]
df_groups_members["group_member_diff"] = df_groups_members["group_members"] - df_groups_members["group_original_members"]
fig = px.histogram(df_groups_members, x="group_member_diff", color="idol_gender")
fig

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



**Insight:** 
* We know from earlier that the mean number of members deceases over time
* Female groups have a skew of gaining 1 or 2 more members
* Female group also have a skew of losing 2 members
* Most male groups lose 1 member at most, with a slight skew of losing 3 members as well


Checking the mean number of idols per male and female group:

In [38]:
print("The mean female group size is:")
df['group_members'][df['idol_gender'] == 'F'].mean()

The mean female group size is:


5.824146981627297

In [39]:
print("The mean male group size is:")
df['group_members'][df['idol_gender'] == 'M'].mean()

The mean male group size is:


7.670967741935484

**Insight:** Male groups have a larger number of members on average.

Checking which artist have the most music videos:

In [40]:
df_music_videos_counts = df_music_videos.groupby(['mv_artist']).size().reset_index(name='counts').sort_values(by=['counts'],ascending=False)
fig = px.bar(df_music_videos_counts, x='mv_artist', y='counts').update_xaxes(categoryorder="total descending")
fig

**Insight:** FTISLAND overwhelmingly have the highest number of videos.

Checking the mean number of videos per idol/group:

In [41]:
print("The mean videos per group is:")
df_music_videos_counts["counts"].mean()

The mean videos per group is:


4.406542056074766

**Insight:** The mean number of videos per idol/group is 4.4.

Checking the most common name for a music video:

In [42]:
df_music_videos["mv_song_name"].value_counts().nlargest(10)

Hello          12
Lonely          7
Paradise        7
Without You     7
Angel           6
Why             6
Tonight         6
Goodbye         6
Beautiful       6
I Love You      5
Name: mv_song_name, dtype: int64

**Insight:** Hello is the most common music video name.

Checking in which month were most music videos released in:

In [43]:
df_mv_release_count = df_music_videos.copy()
df_mv_release_count["mv_date"] = df_music_videos["mv_date"].apply(pd.to_datetime)

df_mv_release_count["mv_date"] = df_mv_release_count["mv_date"].dt.month
df_mv_release_count = df_mv_release_count.groupby(['mv_date', 'mv_type']).size().reset_index(name='counts')

df_mv_release_count['mv_date_month'] = df_mv_release_count['mv_date'].apply(lambda x: calendar.month_abbr[x])
df_mv_release_count

mv_date    mv_type  counts mv_date_month
0         1        Boy     177           Jan
1         1   Boy Solo      19           Jan
2         1      Co-ed      10           Jan
3         1       Girl     116           Jan
4         1  Girl Solo      25           Jan
..      ...        ...     ...           ...
58       12        Boy     114           Dec
59       12   Boy Solo       7           Dec
60       12      Co-ed      14           Dec
61       12       Girl      68           Dec
62       12  Girl Solo      12           Dec

[63 rows x 4 columns]

In [44]:
fig = px.bar(df_mv_release_count, x='mv_date_month', y='counts', color="mv_type").update_xaxes(categoryorder='array', categoryarray= ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'])
fig

**Insight:** The correlation of music video release month is similar to the months in which groups had debuted in.

The K-Pop generations are loosely defined by the following years:
* Gen 1 = 1996-2004
* Gen 2 = 2005-2012
* Gen 3 = 2013-2019
* Gen 4 = 2019-Present

K-Pop generations source obtained from [here](https://www.creatrip.com/en/blog/10988).

Checking if music video release month differs by K-Pop generation:

In [45]:
# convert date string column to datetime format
df_music_videos["mv_date"] = df_music_videos["mv_date"].apply(pd.to_datetime)
df_music_videos.dtypes


mv_date           datetime64[ns]
mv_artist                 object
mv_song_name              object
mv_korean_name            object
mv_director               object
mv_video_url              object
mv_type                   object
mv_release                object
dtype: object

In [46]:
# add month and year columns to existing dataframe
df_music_videos["mv_date_year"] = df_music_videos["mv_date"].dt.year
df_music_videos["mv_date_month"] = df_music_videos["mv_date"].dt.month
df_music_videos


mv_date           mv_artist  ... mv_date_year mv_date_month
0    2020-05-22             Agust D  ...         2020             5
1    2020-05-21               Yubin  ...         2020             5
2    2020-05-21           OnlyOneOf  ...         2020             5
3    2020-05-20         Ryu Sujeong  ...         2020             5
4    2020-05-20               Crush  ...         2020             5
...         ...                 ...  ...          ...           ...
3767 1994-08-08  Seo Taiji and Boys  ...         1994             8
3768 1993-06-06  Seo Taiji and Boys  ...         1993             6
3769 1993-06-06  Seo Taiji and Boys  ...         1993             6
3770 1992-03-23  Seo Taiji and Boys  ...         1992             3
3771 1992-03-23  Seo Taiji and Boys  ...         1992             3

[3772 rows x 10 columns]

In [47]:
# use lambda function to add the generation type to a new column defnied by the music video date (year)
df_music_videos["mv_date_generation"] = df_music_videos['mv_date_year'].apply(lambda x: 'Generation 1' if x <= 2004 else ('Generation 2' if x <= 2012 else ('Generation 3' if x <= 2019 else 'Generation 4')))
df_music_videos

mv_date           mv_artist  ... mv_date_month mv_date_generation
0    2020-05-22             Agust D  ...             5       Generation 4
1    2020-05-21               Yubin  ...             5       Generation 4
2    2020-05-21           OnlyOneOf  ...             5       Generation 4
3    2020-05-20         Ryu Sujeong  ...             5       Generation 4
4    2020-05-20               Crush  ...             5       Generation 4
...         ...                 ...  ...           ...                ...
3767 1994-08-08  Seo Taiji and Boys  ...             8       Generation 1
3768 1993-06-06  Seo Taiji and Boys  ...             6       Generation 1
3769 1993-06-06  Seo Taiji and Boys  ...             6       Generation 1
3770 1992-03-23  Seo Taiji and Boys  ...             3       Generation 1
3771 1992-03-23  Seo Taiji and Boys  ...             3       Generation 1

[3772 rows x 11 columns]

In [48]:
df_music_videos["mv_date_generation"].value_counts()

Generation 3    2753
Generation 2     775
Generation 4     167
Generation 1      77
Name: mv_date_generation, dtype: int64

**Insight:** Generation 3 accounts for 72.98% of all music videos despite a 6 year span.

In [49]:
df_mv_release_count_by_generation = df_music_videos.groupby(['mv_date_generation', 'mv_date_month']).size().reset_index(name='counts')
df_mv_release_count_by_generation['mv_date_month_name'] = df_mv_release_count_by_generation['mv_date_month'].apply(lambda x: calendar.month_abbr[x])
df_mv_release_count_by_generation

mv_date_generation  mv_date_month  counts mv_date_month_name
0        Generation 1              1      10                Jan
1        Generation 1              2       1                Feb
2        Generation 1              3       5                Mar
3        Generation 1              4       4                Apr
4        Generation 1              5       7                May
5        Generation 1              6       4                Jun
6        Generation 1              7       4                Jul
7        Generation 1              8       8                Aug
8        Generation 1              9       2                Sep
9        Generation 1             10      15                Oct
10       Generation 1             11       9                Nov
11       Generation 1             12       8                Dec
12       Generation 2              1      50                Jan
13       Generation 2              2      41                Feb
14       Generation 2              3      56                Mar
15       Generation 2              4      63                Apr
16       Generation 2              5      75                May
17       Generation 2              6      65                Jun
18       Generation 2              7      72                Jul
19       Generation 2              8      72                Aug
20       Generation 2              9      74                Sep
21       Generation 2             10      82                Oct
22       Generation 2             11      69                Nov
23       Generation 2             12      56                Dec
24       Generation 3              1     250                Jan
25       Generation 3              2     195                Feb
26       Generation 3              3     221                Mar
27       Generation 3              4     249                Apr
28       Generation 3              5     229                May
29       Generation 3              6     226                Jun
30       Generation 3              7     280                Jul
31       Generation 3              8     252                Aug
32       Generation 3              9     191                Sep
33       Generation 3             10     239                Oct
34       Generation 3             11     270                Nov
35       Generation 3             12     151                Dec
36       Generation 4              1      37                Jan
37       Generation 4              2      40                Feb
38       Generation 4              3      27                Mar
39       Generation 4              4      36                Apr
40       Generation 4              5      27                May

In [50]:
fig = px.bar(df_mv_release_count_by_generation[df_mv_release_count_by_generation["mv_date_generation"] == "Generation 4"], x='mv_date_month_name', y='counts').update_xaxes(categoryorder='array', categoryarray= ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'])
fig

**Insight:** Generation 4 (current) does not show any obvious signs of having a different trend in release month.

Comparing each generation as a YoY time series: 

In [51]:
fig = px.line(df_mv_release_count_by_generation, x="mv_date_month_name", y="counts", color='mv_date_generation', text="counts").update_traces(textposition="bottom right")
fig

**Insight:** 
* Generations 1 & 2 trending very similarly over time
* Generation 3 has a more signification increase for the July, and November monmths
* Generation 3 has a larger drop for the months of Feburary and December